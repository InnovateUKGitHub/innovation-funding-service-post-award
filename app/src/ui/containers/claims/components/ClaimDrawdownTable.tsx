import { LoanDto, ProjectDto } from "@framework/dtos";
import { checkProjectCompetition } from "@ui/helpers/check-competition-type";
import { getPending } from "@ui/helpers/get-pending";
import { useStores } from "@ui/redux";

import * as ACC from "@ui/components";

type ProjectOptions = Pick<ProjectDto, "id" | "competitionType">;

export interface ClaimDrawdownTableProps extends ProjectOptions {
  /** @description Current project period number could be ahead/behind of the claim period  */
  requiredPeriod: number;
}

function useClaimDrawdown(props: ClaimDrawdownTableProps) {
  const { isLoans } = checkProjectCompetition(props.competitionType);
  const stores = useStores();

  // Note: Only drawdowns can display drawdown data
  if (!isLoans) return null;

  const loanQuery = stores.loans.get(props.id, undefined, props.requiredPeriod);
  const { error, ...rest } = getPending(loanQuery);

  if (error?.message === "No loan found.") return null;

  if (rest.isRejected) throw Error(error?.message ?? "There was an error fetching data within ClaimDrawdownTable");

  // Note: We require total values for the UI
  if (rest.payload && rest.payload.totals === undefined) throw Error("Loan totals must be available.");

  const isLoading = !rest.payload || rest.isLoading;

  return {
    ...rest,
    isLoading,
  };
}

export function ClaimDrawdownTable(props: ClaimDrawdownTableProps) {
  const claimDrawdown = useClaimDrawdown(props);

  if (!claimDrawdown) return null;

  if (claimDrawdown.isLoading) return <ACC.LoadingMessage />;

  const PeriodDrawdownTable = ACC.TypedTable<Required<LoanDto>>();

  return (
    <PeriodDrawdownTable.Table qa="period-loan-table" data={[claimDrawdown.payload as Required<LoanDto>]}>
      <PeriodDrawdownTable.Custom header="Current drawdown" qa="drawdown-id" value={x => x.period} />

      <PeriodDrawdownTable.Currency header="Total loan amount" qa="drawdown-amount" value={x => x.totals.totalLoan} />

      <PeriodDrawdownTable.Currency
        header="Drawdowns to date"
        qa="drawdown-totals-todate"
        value={x => x.totals.totalPaidToDate}
      />

      <PeriodDrawdownTable.Currency header="Drawdown this period" qa="drawdown-current-amount" value={x => x.amount} />

      <PeriodDrawdownTable.Currency
        header="Remaining loan amount"
        qa="drawdown-remaining-amount"
        value={x => x.totals.remainingLoan}
      />
    </PeriodDrawdownTable.Table>
  );
}
