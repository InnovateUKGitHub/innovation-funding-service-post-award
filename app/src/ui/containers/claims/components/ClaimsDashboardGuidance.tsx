import { PartnerDto } from "@framework/dtos";
import { getAuthRoles, ProjectRole } from "@framework/types";
import { checkProjectCompetition } from "@ui/helpers/check-competition-type";
import { useContent } from "@ui/hooks";

import * as Acc from "@ui/components";
import { Content, EmailContent } from "@ui/components";

export interface ClaimsDashboardGuidanceProps {
  overdueProject: PartnerDto["overdueProject"];
  competitionType?: PartnerDto["competitionType"];
  roles?: PartnerDto["roles"];
}

export const ClaimsDashboardGuidance = ({
  overdueProject,
  competitionType = "",
  roles = ProjectRole.Unknown,
}: ClaimsDashboardGuidanceProps) => {
  const { getContent } = useContent();

  let guidanceMessage: React.ReactElement;

  if (overdueProject) {
    guidanceMessage = (
      <Content
        value={x => x.claimsMessages.overdueGuidanceMessage.message}
        components={[<EmailContent key="email" value={x => x.claimsMessages.overdueGuidanceMessage.email} />]}
      />
    );
  } else {
    const isPartnerFc = getAuthRoles(roles).isFc;
    const { isCombinationOfSBRI } = checkProjectCompetition(competitionType);

    const defaultGuidanceMessage = getContent(x => x.claimsMessages.guidanceMessage);

    const displaySbriGuidance = isPartnerFc && isCombinationOfSBRI;

    if (displaySbriGuidance) {
      return <Acc.Renderers.SimpleString>{defaultGuidanceMessage}</Acc.Renderers.SimpleString>;
    }

    guidanceMessage = <Acc.Renderers.Markdown trusted value={defaultGuidanceMessage} />;
  }

  return <Acc.ValidationMessage qa="guidance-message" messageType="info" message={guidanceMessage} />;
};
