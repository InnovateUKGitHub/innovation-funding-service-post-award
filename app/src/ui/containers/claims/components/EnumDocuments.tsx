import { DocumentDescription } from "@framework/constants/documentDescription";
import { DropdownOption } from "@ui/components/form";
import React from "react";
import { useEnumDocuments } from "./allowed-documents.hook";

interface EnumDocumentsProps {
  documentsToCheck: Readonly<DocumentDescription[]>;
  children: (documents: DropdownOption[]) => React.ReactElement;
}

/**
 * wrapper component to allow use of `useEnumDocuments` hook with class components
 */
export function EnumDocuments({ documentsToCheck, children }: EnumDocumentsProps) {
  const documents = useEnumDocuments(DocumentDescription, documentsToCheck);

  return children(documents);
}
