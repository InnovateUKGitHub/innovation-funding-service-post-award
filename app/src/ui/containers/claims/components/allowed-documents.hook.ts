import { DocumentDescription, getDocumentDescriptionContentSelector } from "@framework/constants";
import { getAllEnumValues } from "@shared/enumHelper";
import { DropdownOption } from "@ui/components";
import { useContent } from "@ui/hooks";

export function useEnumDocuments(
  enumDocuments: object,
  documentsToCheck: Readonly<DocumentDescription[]>,
): DropdownOption[] {
  const { getContent } = useContent();
  const getRawDocs: number[] = getAllEnumValues(enumDocuments);

  return getRawDocs.reduce<DropdownOption[]>((acc, doc) => {
    const isInvalidDocument = !documentsToCheck.includes(doc);

    if (isInvalidDocument) return acc;

    const id = doc.toString();
    const value = getContent(getDocumentDescriptionContentSelector(doc));

    return [
      ...acc,
      {
        id,
        value,
      },
    ];
  }, []);
}
