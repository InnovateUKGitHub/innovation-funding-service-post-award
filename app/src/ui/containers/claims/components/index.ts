export * from "./ClaimsDashboardGuidance";
export * from "./EnumDocuments";
export * from "./allowed-documents.hook";
