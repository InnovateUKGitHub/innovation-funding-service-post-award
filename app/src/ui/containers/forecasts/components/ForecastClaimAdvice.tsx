import { useContent } from "@ui/hooks";
import { ILinkInfo } from "@framework/types";
import { SimpleString } from "@ui/components/renderers";
import { Link } from "@ui/components";

export interface ForecastClaimAdviceProps {
  claimLink: ILinkInfo;
}

export function ForecastClaimAdvice({ claimLink }: ForecastClaimAdviceProps) {
  const { getContent } = useContent();

  const part1 = getContent(x => x.pages.forecastsComponents.adviseMessage.part1);
  const part2Link = getContent(x => x.pages.forecastsComponents.adviseMessage.part2Link);
  const part3 = getContent(x => x.pages.forecastsComponents.adviseMessage.part3);

  return (
    <SimpleString>
      {part1}
      <Link route={claimLink}>{part2Link}</Link>
      {part3}
    </SimpleString>
  );
}
