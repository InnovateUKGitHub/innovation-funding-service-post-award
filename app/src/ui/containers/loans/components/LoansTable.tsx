import cx from "classnames";
import { LoanStatus } from "@framework/entities";
import { ILinkInfo } from "@framework/types";
import * as ACC from "@ui/components";
import type { Loan } from "../loanOverview.logic";
import { useContent } from "@ui/hooks";
import { ReactNode } from "react";

export interface LoansTableProps {
  items: Loan[];
  createLink: (id: LoanId) => ILinkInfo;
  roles: { readonly isMo: boolean; readonly isFc: boolean; readonly isPm: boolean };
}

const Drawdown = ACC.createTypedTable<Loan>();
export const LoansTable = ({ items, createLink, roles }: LoansTableProps) => {
  const { getContent } = useContent();

  const nextLoanIndex = items.findIndex(x => x.status === LoanStatus.PLANNED);
  const isFirstLoanRequest = nextLoanIndex === 0;
  const nextLoan = items[nextLoanIndex];
  const previousLoanItem = items[nextLoanIndex - 1];

  const hasPreviousApprovedLoan = previousLoanItem?.status === LoanStatus.APPROVED;

  // Note: The first request will always be available (when PLANNED)
  const canRequestLoan = isFirstLoanRequest ? true : hasPreviousApprovedLoan;

  const getRowClassName = (loan: typeof items[0]) => {
    const isAvailable = nextLoan?.id === loan.id;
    const isLocked = loan.status === LoanStatus.APPROVED;
    const isUnknown = loan.status === LoanStatus.UNKNOWN;

    return cx("loan-table_row", {
      "loan-table_row--available": !isUnknown && isAvailable,
      "loan-table_row--unavailable": !isUnknown && !isLocked && !isAvailable,
      "loan-table_row--locked": isUnknown || (isLocked && !isAvailable),
    });
  };

  return (
    <Drawdown.Table data={items} qa="drawdown-list" className="loan-table" bodyRowClass={getRowClassName}>
      <Drawdown.String
        header={getContent(x => x.components.loansTable.drawdown)}
        qa="drawdown-period"
        value={x => `${x.period}`}
      />

      <Drawdown.FullNumericDate
        header={getContent(x => x.components.loansTable.dueDate)}
        qa="drawdown-date"
        value={x => x.requestDate}
      />

      <Drawdown.Custom
        header={getContent(x => x.components.loansTable.drawdownAmount)}
        qa="drawdown-forecast-amount"
        classSuffix="numeric"
        value={x => (
          <ACC.Renderers.Bold>
            <ACC.Renderers.Currency fractionDigits={0} value={x.forecastAmount} />
          </ACC.Renderers.Bold>
        )}
      />

      <Drawdown.String
        header={getContent(x => x.components.loansTable.status)}
        qa="drawdown-status"
        value={x => x.status}
      />

      <Drawdown.Custom
        qa="drawdown-action"
        value={({ id }) => {
          const nextPlannedLoan = nextLoan?.id === id;

          return (
            nextPlannedLoan &&
            !roles.isMo && (
              <LoanRequestButton route={createLink(nextLoan.id)} disabled={!canRequestLoan}>
                {roles.isFc && getContent(x => x.components.loansTable.request)}
                {roles.isPm && getContent(x => x.components.loansTable.view)}
              </LoanRequestButton>
            )
          );
        }}
      />
    </Drawdown.Table>
  );
};

interface LoanRequestButtonProps {
  disabled: boolean;
  route: ILinkInfo;
  children: ReactNode;
}

const LoanRequestButton = ({ disabled, route, children }: LoanRequestButtonProps) => {
  return disabled ? (
    <ACC.Button disabled styling="Primary" className="govuk-!-margin-bottom-0">
      {children}
    </ACC.Button>
  ) : (
    <ACC.Link route={route} styling="PrimaryButton" className="govuk-!-margin-bottom-0">
      {children}
    </ACC.Link>
  );
};
