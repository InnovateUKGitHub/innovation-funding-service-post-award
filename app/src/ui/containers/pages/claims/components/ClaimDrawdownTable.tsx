import { LoanDto } from "@framework/dtos/loanDto";
import { ProjectDto } from "@framework/dtos/projectDto";
import { LoadingMessage } from "@ui/components/bjss/loading";
import { createTypedTable } from "@ui/components/atomicDesign/molecules/Table/Table";
import { checkProjectCompetition } from "@ui/helpers/check-competition-type";
import { getPending } from "@ui/helpers/get-pending";
import { useStores } from "@ui/redux/storesProvider";

type ProjectOptions = Pick<ProjectDto, "id" | "competitionType">;

export interface ClaimDrawdownTableProps extends ProjectOptions {
  /** @description Current project period number could be ahead/behind of the claim period  */
  requiredPeriod: number;
}

const PeriodDrawdownTable = createTypedTable<Required<LoanDto>>();

const useClaimDrawdown = (props: ClaimDrawdownTableProps) => {
  const { isLoans } = checkProjectCompetition(props.competitionType);
  const stores = useStores();

  // Note: Only drawdowns can display drawdown data
  if (!isLoans) return null;

  const loanQuery = stores.loans.get(props.id, undefined, props.requiredPeriod);
  const { error, ...rest } = getPending(loanQuery);

  if (error?.message === "No loan found.") return null;

  if (rest.isRejected) throw Error(error?.message ?? "There was an error fetching data within ClaimDrawdownTable");

  // Note: We require total values for the UI
  if (rest.payload && rest.payload.totals === undefined) throw Error("Loan totals must be available.");

  const isLoading = !rest.payload || rest.isLoading;

  return {
    ...rest,
    isLoading,
  };
};

export const ClaimDrawdownTable = (props: ClaimDrawdownTableProps) => {
  const claimDrawdown = useClaimDrawdown(props);

  if (!claimDrawdown) return null;

  if (claimDrawdown.isLoading) return <LoadingMessage />;

  return (
    <PeriodDrawdownTable.Table qa="period-loan-table" data={[claimDrawdown.payload as Required<LoanDto>]}>
      <PeriodDrawdownTable.Custom header="Current drawdown" qa="drawdown-id" value={x => x.period} />

      <PeriodDrawdownTable.Currency header="Total loan amount" qa="drawdown-amount" value={x => x.totals.totalLoan} />

      <PeriodDrawdownTable.Currency
        header="Drawdowns to date"
        qa="drawdown-totals-todate"
        value={x => x.totals.totalPaidToDate}
      />

      <PeriodDrawdownTable.Currency header="Drawdown this period" qa="drawdown-current-amount" value={x => x.amount} />

      <PeriodDrawdownTable.Currency
        header="Remaining loan amount"
        qa="drawdown-remaining-amount"
        value={x => x.totals.remainingLoan}
      />
    </PeriodDrawdownTable.Table>
  );
};
