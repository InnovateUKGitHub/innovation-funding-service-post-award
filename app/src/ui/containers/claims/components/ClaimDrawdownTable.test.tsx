import { render } from "@testing-library/react";

import { TestBed, TestBedContent, TestBedStore } from "@shared/TestBed";
import { ClaimDrawdownTable, ClaimDrawdownTableProps } from "@ui/containers/claims/components/ClaimDrawdownTable";
import { LoanDto } from "@framework/dtos";
import { LoanStatus } from "@framework/entities";
import { createProjectDto } from "@framework/util/stubDtos";
import { Pending } from "@shared/pending";
import { LoadingStatus } from "@framework/constants";

describe("<ClaimDrawdownTable />", () => {
  const stubProject = createProjectDto({ competitionType: "LOANS" });

  const stubLoan: LoanDto = {
    id: "a0H3M0000002GRYUA2",
    status: LoanStatus.APPROVED,
    period: 1,
    requestDate: new Date(Date.UTC(1, 10)),
    amount: 10000,
    forecastAmount: 10000,
    comments: "",
    totals: {
      totalLoan: 30000,
      totalPaidToDate: 6000,
      remainingLoan: 14000,
    },
  };

  const defaultProps: ClaimDrawdownTableProps = {
    competitionType: stubProject.competitionType,
    id: stubProject.id,
    requiredPeriod: stubLoan.period,
  };

  const stubContent = {
    components: {
      loading: {
        message: { content: "stub-loading" },
      },
    },
  };

  const setup = (props?: Partial<ClaimDrawdownTableProps>, stubLoanQuery: Pending<any> = Pending.done(stubLoan)) => {
    const stubStore = {
      loans: {
        get: jest.fn().mockReturnValue(stubLoanQuery),
      },
    } as any;

    return render(
      <TestBed stores={stubStore as TestBedStore} content={stubContent as TestBedContent}>
        <ClaimDrawdownTable {...defaultProps} {...props} />
      </TestBed>,
    );
  };

  describe("@returns", () => {
    test("with no UI when compeition is not LOANS", () => {
      const { container } = setup({ competitionType: "CR&D" });

      expect(container.firstChild).toBeNull();
    });

    test("with UI when compeition is LOANS", () => {
      const validQuery = Pending.done(stubLoan);
      const { queryByTestId } = setup({ competitionType: "LOANS" }, validQuery);

      const claimDrawdownTable = queryByTestId("period-loan-table");

      expect(claimDrawdownTable).toMatchSnapshot();
    });

    test("with loading state", () => {
      const loadingQuery = new Pending(LoadingStatus.Loading, { ...stubLoan, totals: undefined });
      const { queryByTestId } = setup(undefined, loadingQuery);

      const loadingUI = queryByTestId("loading-message");

      expect(loadingUI).toBeInTheDocument();
    });

    test("when requiredPeriod could not be found", () => {
      const noLoanFoundQuery = new Pending(LoadingStatus.Failed, undefined, {
        message: "No loan found.",
      });
      const { container } = setup(undefined, noLoanFoundQuery);

      expect(container.firstChild).toBeNull();
    });

    describe("with an error", () => {
      // TODO: Figure out a better way to silencing console errors for thrown expect's...
      jest.spyOn(console, "error").mockImplementation(jest.fn);

      test("with no totals provided from query", () => {
        const misingTotalsQuery = Pending.done({ ...stubLoan, totals: undefined });

        expect(() => setup(undefined, misingTotalsQuery)).toThrow("Loan totals must be available.");
      });

      test("with a default error", () => {
        const errorQuery = new Pending(LoadingStatus.Failed);

        expect(() => setup(undefined, errorQuery)).toThrow(
          "There was an error fetching data within ClaimDrawdownTable",
        );
      });

      test("with a thrown message", () => {
        const stubErrorMessage = "stub-error";
        const errorQuery = new Pending(LoadingStatus.Failed, undefined, { message: stubErrorMessage });

        expect(() => setup(undefined, errorQuery)).toThrow(stubErrorMessage);
      });

      test("with a throw error no message", () => {
        const errorQuery = new Pending(LoadingStatus.Failed, undefined, {});

        expect(() => setup(undefined, errorQuery)).toThrow(
          "There was an error fetching data within ClaimDrawdownTable",
        );
      });
    });
  });
});
