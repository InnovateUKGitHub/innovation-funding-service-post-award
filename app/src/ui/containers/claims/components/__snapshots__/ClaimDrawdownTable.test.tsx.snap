// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ClaimDrawdownTable /> @returns with UI when competition is LOANS 1`] = `
<div
  class="govuk-table-wrapper"
  data-qa="period-loan-table"
>
  <table
    class="govuk-table"
  >
    <colgroup>
      <col
        data-qa="col-drawdown-id"
      />
      <col
        data-qa="col-drawdown-amount"
      />
      <col
        data-qa="col-drawdown-totals-todate"
      />
      <col
        data-qa="col-drawdown-current-amount"
      />
      <col
        data-qa="col-drawdown-remaining-amount"
      />
    </colgroup>
    <thead
      class="govuk-table__head"
    >
      <tr
        class="govuk-table__row"
      >
        <th
          class="govuk-table__header"
          scope="col"
        >
          Current drawdown
        </th>
        <th
          class="govuk-table__header govuk-table__header--numeric"
          scope="col"
        >
          Total loan amount
        </th>
        <th
          class="govuk-table__header govuk-table__header--numeric"
          scope="col"
        >
          Drawdowns to date
        </th>
        <th
          class="govuk-table__header govuk-table__header--numeric"
          scope="col"
        >
          Drawdown this period
        </th>
        <th
          class="govuk-table__header govuk-table__header--numeric"
          scope="col"
        >
          Remaining loan amount
        </th>
      </tr>
    </thead>
    <tbody
      class="govuk-table__body"
    >
      <tr
        class="govuk-table__row"
      >
        <td
          class="govuk-table__cell"
        >
          1
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          <span
            class="currency"
            style="white-space: nowrap;"
          >
            £30
            ,
            <wbr />
            000.00
          </span>
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          <span
            class="currency"
            style="white-space: nowrap;"
          >
            £6
            ,
            <wbr />
            000.00
          </span>
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          <span
            class="currency"
            style="white-space: nowrap;"
          >
            £10
            ,
            <wbr />
            000.00
          </span>
        </td>
        <td
          class="govuk-table__cell govuk-table__cell--numeric"
        >
          <span
            class="currency"
            style="white-space: nowrap;"
          >
            £14
            ,
            <wbr />
            000.00
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
